File: Astromaximum/Classes/utils/AmaxLocationBundle.m
//
//  AmaxLocationBundle.m
//  Astromaximum
//
//  Created by admin on 14.11.12.
//  Copyright (c) 2012 S&W Axis. All rights reserved.
//

#import "Astromaximum-Bridging-Header.h"
#import "Astromaximum-Swift.h"

@implementation AmaxLocationBundle

@synthesize mRecordCount = _mRecordCount;

- (id)initWithFilePath:(NSString *)filePath
{
    NSData *fullData = [NSData dataWithContentsOfFile:filePath];
    AmaxDataInputStream *is = [[AmaxDataInputStream alloc]initWithData:fullData];
    [is readShort]; // skip year
    _mRecordCount = [is readUnsignedShort];
    recordLengths = malloc(_mRecordCount * sizeof(recordLengths[0]));
    for (int i = 0; i < _mRecordCount; ++i)
        recordLengths[i] = [is readUnsignedShort];
    Size bufferLength = [is availableBytes];
    NSData* locData = [is readWithLength:bufferLength];
    locStream = [[AmaxDataInputStream alloc]initWithData:locData];
    return self;
}

- (void) dealloc
{
    free(recordLengths);
}

- (Size) extractLocationByIndex:(int)index intoData:(NSData *)data
{
    [locStream reset];
    Size offset = 0;
    for (int i = 0; i < index; i++) {
        offset += recordLengths[i];
    }
    Size length = recordLengths[index];
    [locStream skipBytes:offset];
    data = [locStream readWithLength:length];
    return length;
}

@end

File: Astromaximum/Classes/utils/AmaxLocationBundle.swift
//
//  AmaxLocationBundle.swift
//  Astromaximum2
//
//  Created by Andrei Ivushkin on 12.07.2020.
//  Copyright © 2020 S&W Axis. All rights reserved.
//

import Foundation

@objcMembers class AmaxLocationBundle : NSObject {
    var recordLengths = [Int]()
    var locStream: AmaxDataInputStream

    init(filePath: String?) {
        let fullData = (NSData(contentsOfFile: filePath ?? "") as Data?)!
        let ais = AmaxDataInputStream(data: fullData)
        _ = ais.readShort() // skip year
        let recordCount = Int(ais.readUnsignedShort())
        for _ in 0..<recordCount {
            recordLengths.append(Int(ais.readUnsignedShort()))
        }
        let bufferLength = ais.availableBytes()
        let buffer = ais.read(length: bufferLength)
        locStream = AmaxDataInputStream(data: buffer)
    }

    func extractLocation(by index: Int) -> Data {
        locStream.reset()
        var offset = 0
        for i in 0 ..< index {
            offset += recordLengths[i]
        }
        let length = recordLengths[index]
        locStream.skipBytes(offset)
        let data = locStream.read(length: length)
        return data
    }
}
